File: legacy/unreal-5.7/Source/KyndeBlade/Combat/CombatAction.cpp
#include "CombatAction.h"
#include "../Characters/MedievalCharacter.h"

UCombatAction::UCombatAction()
{
	ActionData = FCombatActionData();
}

void UCombatAction::ExecuteAction(AMedievalCharacter* Executor, AMedievalCharacter* Target)
{
	if (!CanExecute(Executor))
	{
		return;
	}

	// Expedition 33-inspired: Check Kynde cost before executing
	if (ActionData.KyndeCost > 0.0f)
	{
		if (!Executor->ConsumeKynde(ActionData.KyndeCost))
		{
			return; // Not enough Kynde, action fails
		}
	}

	if (Executor)
	{
		Executor->ConsumeStamina(ActionData.StaminaCost);
		if (ActionData.ManaCost > 0.0f)
		{
			Executor->ConsumeMana(ActionData.ManaCost);
		}
	}

	// Expedition 33-inspired: Generate Kynde from melee attacks
	if (ActionData.KyndeGenerated > 0.0f && Executor)
	{
		Executor->GainKynde(ActionData.KyndeGenerated);
	}

	// Execute the action based on type (Piers Plowman terminology)
	switch (ActionData.ActionType)
	{
	case ECombatActionType::Strike: // Attack - "Sothfastnesse's Stroke"
		if (Target)
		{
			float FinalDamage = ActionData.Damage;

			// Expedition 33-inspired: Broken enemies take 50% more damage
			if (Target->IsBroken())
			{
				FinalDamage *= 1.5f;
			}

			Target->ApplyCustomDamage(FinalDamage, Executor);

			// Expedition 33-inspired: Apply break damage
			if (ActionData.BreakDamage > 0.0f && Target)
			{
				Target->TakeBreakDamage(ActionData.BreakDamage);
			}
		}
		break;
	case ECombatActionType::Escapade: // Dodge - "Kynde's Evasion"
		// Escapade logic handled in real-time
		break;
	case ECombatActionType::Ward: // Parry - "Trewthe's Sheeld"
		// Ward logic handled in real-time
		break;
	case ECombatActionType::Counter: // Counter-attack
		if (Target)
		{
			Target->ApplyCustomDamage(ActionData.Damage * 1.5f, Executor);
		}
		break;
	case ECombatActionType::Guard:
		if (Executor)
		{
			Executor->SetGuarding(true);
			Executor->RestoreStamina(15.0f);
		}
		break;
	case ECombatActionType::Rest: // Wait - "Kynde's Rest"
		if (Executor)
		{
			Executor->RestoreStamina(20.0f);
		}
		break;
	default:
		break;
	}

	OnActionExecuted(Executor, Target);
}

void UCombatAction::ExecuteActionWithDeferredDamage(AMedievalCharacter* Executor, AMedievalCharacter* Target)
{
	if (!CanExecute(Executor))
	{
		return;
	}

	if (ActionData.KyndeCost > 0.0f)
	{
		if (!Executor->ConsumeKynde(ActionData.KyndeCost))
		{
			return;
		}
	}

	if (Executor)
	{
		Executor->ConsumeStamina(ActionData.StaminaCost);
		if (ActionData.ManaCost > 0.0f)
		{
			Executor->ConsumeMana(ActionData.ManaCost);
		}
	}

	if (ActionData.KyndeGenerated > 0.0f && Executor)
	{
		Executor->GainKynde(ActionData.KyndeGenerated);
	}

	switch (ActionData.ActionType)
	{
	case ECombatActionType::Strike:
		// Damage and break applied by TurnManager after real-time window
		break;
	case ECombatActionType::Escapade:
	case ECombatActionType::Ward:
		break;
	case ECombatActionType::Counter:
		if (Target)
		{
			Target->ApplyCustomDamage(ActionData.Damage * 1.5f, Executor);
		}
		break;
	case ECombatActionType::Guard:
		if (Executor)
		{
			Executor->SetGuarding(true);
			Executor->RestoreStamina(15.0f);
		}
		break;
	case ECombatActionType::Rest:
		if (Executor)
		{
			Executor->RestoreStamina(20.0f);
		}
		break;
	default:
		break;
	}

	OnActionExecuted(Executor, Target);
}

bool UCombatAction::CanExecute(AMedievalCharacter* Executor)
{
	if (!Executor)
	{
		return false;
	}

	if (Executor->GetCurrentStamina() < ActionData.StaminaCost)
	{
		return false;
	}
	if (ActionData.ManaCost > 0.0f && Executor->GetCurrentMana() < ActionData.ManaCost)
	{
		return false;
	}
	return true;
}

File: KyndeBlade_Unity/Assets/KyndeBlade/Code/Combat/CombatAction.cs
using System;
using UnityEngine;
using KyndeBlade;

namespace KyndeBlade.Combat
{
    public enum CombatActionType
    {
        Strike,
        Escapade,
        Ward,
        Counter,
        Special,
        Rest,
        RangedStrike,
        Heal
    }

    public enum KyndeElementType
    {
        None,
        Flamme,
        Frost,
        Thunder,
        Trewthe,
        Fals,
        Kynde
    }

    [Serializable]
    public class CombatActionData
    {
        public CombatActionType ActionType = CombatActionType.Rest;
        public float Damage;
        public float StaminaCost;
        public float SuccessWindow;
        public float CastTime;
        public float ExecutionTime;
        public float KyndeGenerated;
        public float KyndeCost;
        public float BreakDamage;
        public KyndeElementType ElementType;
        public string ActionName = "Kyndes Rest";
    }

    [CreateAssetMenu(fileName = "NewCombatAction", menuName = "KyndeBlade/Combat Action")]
    public class CombatAction : ScriptableObject
    {
        public CombatActionData ActionData = new CombatActionData();

        public virtual void ExecuteAction(MedievalCharacter executor, MedievalCharacter target)
        {
            if (!CanExecute(executor)) return;

            if (ActionData.KyndeCost > 0f && !executor.ConsumeKynde(ActionData.KyndeCost))
                return;

            executor.ConsumeStamina(ActionData.StaminaCost);

            if (ActionData.KyndeGenerated > 0f)
                executor.GainKynde(ActionData.KyndeGenerated);

            switch (ActionData.ActionType)
            {
                case CombatActionType.Strike:
                case CombatActionType.Counter:
                case CombatActionType.RangedStrike:
                    if (target != null)
                    {
                        if (executor.WillDeferHitToAnimation())
                            executor.RegisterPendingHit(this, target);
                        else
                            ApplyHitEffect(executor, target);
                    }
                    break;
                case CombatActionType.Heal:
                    var healTarget = target != null ? target : executor;
                    if (ActionData.Damage > 0f) healTarget.Heal(ActionData.Damage);
                    healTarget.RemoveHungerStack();
                    break;
                case CombatActionType.Rest:
                    executor.RestoreStamina(20f);
                    executor.RemoveHungerStack();
                    break;
            }

            // For deferred hit (Strike/Counter/RangedStrike), OnActionExecuted is called from OnAnimationHit
            if (ActionData.ActionType != CombatActionType.Strike && ActionData.ActionType != CombatActionType.Counter && ActionData.ActionType != CombatActionType.RangedStrike)
                OnActionExecuted(executor, target);
            else if (!executor.WillDeferHitToAnimation())
                OnActionExecuted(executor, target);
        }

        /// <summary>Apply damage and break for this action. Called immediately or from Animation Event (OnAnimationHit). Override in subclasses for custom hit effects.</summary>
        public virtual void ApplyHitEffect(MedievalCharacter executor, MedievalCharacter target)
        {
            if (target == null) return;
            float dmg = CombatCalculator.CalculateDamage(executor, target, this);
            if (ActionData.ActionType == CombatActionType.Counter) dmg *= 1.5f;
            if (target.IsBroken()) dmg *= GameWorldConstants.BrokenDamageMultiplier;
            var tm = GameRuntime.TurnManager;
            if (tm != null) dmg = tm.ApplyCritToDamage(dmg, out _);
            target.ApplyCustomDamage(dmg, executor, damageAlreadyFinal: true);
            if (ActionData.BreakDamage > 0f) target.TakeBreakDamage(ActionData.BreakDamage);
        }

        /// <summary>Called after hit is applied (e.g. from animation event). Fires OnActionExecuted for observers.</summary>
        public void NotifyHitExecuted(MedievalCharacter executor, MedievalCharacter target)
        {
            OnActionExecuted(executor, target);
        }

        protected virtual void OnActionExecuted(MedievalCharacter executor, MedievalCharacter target) { }

        protected virtual bool CanExecute(MedievalCharacter executor)
        {
            if (executor == null) return false;
            if (executor.GetCurrentStamina() < ActionData.StaminaCost) return false;
            if (ActionData.KyndeCost > 0f && executor.GetCurrentKynde() < ActionData.KyndeCost) return false;
            return true;
        }

        public float GetCastTime() => ActionData.CastTime;
        public float GetExecutionTime() => ActionData.ExecutionTime;
    }
}
